### ISSUE TITLE
Draft a specification?

### ISSUE THREAD

**Issue by jsmorph** [timbray/quamina#210]
I thought I'd create an issue for any on-going discussion:

As we've mentioned [here](https://github.com/timbray/quamina/issues/196#issuecomment-1426533250) and sort of in #57, it might be good to have some sort of specification for the pattern language. The current [documentation](https://github.com/timbray/quamina/blob/main/PATTERNS.md) is of course an excellent source, but adding something more formal is appealing (to me).

Ideally the specification is flexible enough to support something like feature flags ("`shellstyle` not supported" or whatever). Good test cases: Evolution of capabilities here (for example: pre-`exists` and post-`exists` specs) and similarly for [Event Ruler](https://github.com/aws/event-ruler).

A potentially interesting twist is having the specification(s) not tied to specific event rendering languages (like JSON or protobuf), but that goal seems like a stretch with questionable net worth at this point.

Thoughts? Any appetite for something like this?

**Comment by timbray** [comment]
Presumably if we're interested in a stable standards doc, the goal is interoperability with other implementations, of which at the moment there is one, namely aws-event-ruler. So I think that if we were going to do this, we should file an issue over there to alert them and invite them to collaborate. Sound reasonable?

While I would be happy to pitch in and help, I don't have the cycles to lead such an effort.

**Comment by embano1** [comment]
This came up several times in our (AWS) discussions as well, paging @nikp and @baldawar

**Comment by timbray** [comment]
If by "converge on" you mean "start from a copy of and preserve the RFC-like style" then I agree.

**Comment by nikp** [comment]
I've never written one of these before but would be keen to help. CC @jonessha as well :)

**Comment by jsmorph** [comment]
@timbray, what do you suggest for the [drafting format](https://authors.ietf.org/choosing-a-format-and-tools) these days? I'm not current at all, but [drafting in Markdown](https://authors.ietf.org/drafting-in-markdown) might work well enough for us here. My weakly held preference is to start with whatever real markup, boilerplate, etc. from the outset (with content from [`PATTERNS.md`](https://github.com/timbray/quamina/blob/main/PATTERNS.md) rather than retrofit later. What guidance do you have? @embano1 @baldawar @nikp and others, what say you?

I'm happy to start a branch (or whatever) with a crude first draft.

**Comment by timbray** [comment]
Standards-wise, these days I live in the IETF. There, more or less all RFCs are created in either Markdown or XML. The tooling is pretty good either way. Have a look at the setup in https://github.com/ietf-wg-jsonpath/iregexp, if you clone that you get a Makefile and everything. If we have an explicit goal to take this to a standards organization we should pick one sooner rather than later because they all have strong opinions about what input drafts should look like. 

My instinct would be to do it as an IETF internet-draft, because of that tooling, and anyone can post an Internet-draft for free, which gives at least a permanent URL.

Hey AWS guys, is there any relationship with CNCF these days?  I sort of did that for a while but ran out of patience with them. 

Oh, @jsmorph,d to answer your question: Yeah, markdown is fine.

**Comment by timbray** [comment]
Yep, new repo feels appropriate to me, wonder what the AWS folks think.

**Comment by baldawar** [comment]
New repo makes sense. 

For full disclosure, so far I'm only speaking on my own behalf. I need to check if there's any due-process that needs to be followed before I work on behalf of AWS. From past experience, these things can take time so I wouldn't recommend waiting on it. I'd be keen to get started. Happy to create a repo w all permissions & scaffolding based on https://github.com/ietf-wg-jsonpath/iregexp over the weekend.

**Comment by baldawar** [comment]
on 

> Hey AWS guys, is there any relationship with CNCF these days? I sort of did that for a while but ran out of patience with them.

Sorry didn't see this before. I'm not sure at the moment. Let me ask in our internal channels.

**Comment by embano1** [comment]
> Hey AWS guys, is there any relationship with CNCF these days? I sort of did that for a while but ran out of patience with them.

I'm still a bit involved in the CloudEvents side (mostly for the Go and PowerShell SDKs). In general, we do have several streams with the CNCF. Which specific project is this question about?

**Comment by baldawar** [comment]
I've created this shell repo temporarily https://github.com/baldawar/draft-ietf-patterns-protocol/ based on https://github.com/ietf-wg-jsonpath/iregexp/ & https://github.com/martinthomson/i-d-template (neat tool, btw!)

Everyone who I could think be interested should be invited to the repo (cc @longzhang-lz  for context on the mysterious invite). I will move this to an organization as soon as someone can suggest a decent name. I could only come up w patterns / matchers / ruler patterns which are all felt too generic. 

Right now I'm heads down on few work-related gremlins, but I can follow up within AWS on if there's any prescribed guidelines the company requires. I also have a bit of backlog of clean up work around [event-ruler](https://github.com/aws/event-ruler) that I'm desperate to wrap up first, so expecting to take a lot more active role. Until then (targeting Apr End) happy to take care of any minor tasks, administrative duties, and reviews.

**Comment by baldawar** [comment]
I'll be starting to push out some PRs on https://github.com/epml-spec (Event Pattern Matching Language) where I've moved the [repo](https://github.com/epml-spec/draft-ietf-patterns-protocol).
